### PR TITLE
Avoids holding tablet lock while doing split computations

### PR DESCRIPTION
**Pull request by keith-turner**
In addition to avoiding tablet lock for split computations, this PR also avoid redoing split computations when the files have not changed.

### CODE REVIEW

**Comment by keith-turner** [comment]
Ignoring white space changes makes the diffs smaller.

**Comment by keith-turner** [comment]
Ran all ITs matching the pattern `*Split*IT` and that looked good.

```
INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.accumulo.test.LargeSplitRowIT
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 43.505 s - in org.apache.accumulo.test.LargeSplitRowIT
[INFO] Running org.apache.accumulo.test.functional.BatchScanSplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 14.504 s - in org.apache.accumulo.test.functional.BatchScanSplitIT
[INFO] Running org.apache.accumulo.test.functional.SplitRecoveryIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.354 s - in org.apache.accumulo.test.functional.SplitRecoveryIT
[INFO] Running org.apache.accumulo.test.functional.AddSplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.51 s - in org.apache.accumulo.test.functional.AddSplitIT
[INFO] Running org.apache.accumulo.test.functional.DeleteTableDuringSplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 67.48 s - in org.apache.accumulo.test.functional.DeleteTableDuringSplitIT
[INFO] Running org.apache.accumulo.test.functional.DeleteRowsSplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 42.34 s - in org.apache.accumulo.test.functional.DeleteRowsSplitIT
[INFO] Running org.apache.accumulo.test.functional.BulkSplitOptimizationIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 16.016 s - in org.apache.accumulo.test.functional.BulkSplitOptimizationIT
[INFO] Running org.apache.accumulo.test.functional.MetadataSplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 19.841 s - in org.apache.accumulo.test.functional.MetadataSplitIT
[INFO] Running org.apache.accumulo.test.functional.SplitIT
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 92.578 s - in org.apache.accumulo.test.functional.SplitIT
[INFO] Running org.apache.accumulo.test.SplitRecoveryIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.61 s - in org.apache.accumulo.test.SplitRecoveryIT
[INFO] Running org.apache.accumulo.test.ManySplitIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 17.312 s - in org.apache.accumulo.test.ManySplitIT
[INFO] Running org.apache.accumulo.test.SplitCancelsMajCIT
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 68.982 s - in org.apache.accumulo.test.SplitCancelsMajCIT
[INFO] Running org.apache.accumulo.test.MetaSplitIT
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 30.511 s - in org.apache.accumulo.test.MetaSplitIT
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 21, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by keith-turner** [comment]
Realized this change could possibly lead to split starvation.  The change allows more concurrency would means if the split takes a long time to compute and files keep coming in that the tablet may never split because its always recomputing the split point.  I think maybe this can be avoided with a simple change which I will work on shortly.

**Review by dlmarion — APPROVED** [review]
LGTM, a minor comment.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 1]
```diff
@@ -1664,85 +1670,75 @@ private SplitRowSpec findSplitRow(Collection<FileRef> files) {
...
+          return new SplitRowSpec(keys.firstKey(), candidate.getRow());
         }
 
-        log.warn("Cannot split tablet " + extent + " it contains a big row : " + lastRow);
+      }
 
-        sawBigRow = true;
-        timeOfLastMinCWhenBigFreakinRowWasSeen = lastMinorCompactionFinishTime;
-        timeOfLastImportWhenBigFreakinRowWasSeen = lastMapFileImportTime;
+      log.warn("Cannot split tablet " + extent + " it contains a big row : " + lastRow);
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1711]: Definitely should use placeholders whenever possible, for safety. You never know what's going to get logged and incorrectly parsed as part of the format string. We took efforts in 2.0 to fix a bunch of those, and we should try to fix it whenever we see them going forward.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1711]: Gave this a try and it did not work.  Looked into why and it seems the Tablet class is using log4j directly for logging and not slf4j.   The lack of `{}` is a widespread issue in this class so I am not going to address that in this PR.  Hopefully things are better when I merge forward to 2.1

[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:1711]: As long as it's updated during the merge to 2.1, I don't mind 1.10 not being fixed. We spent a lot of time updating these in 2.0, so we don't want to regress on the merge.

**Pull request merged**